### ISSUE TITLE
[question] CMake variable for direct library dependencies only

### ISSUE THREAD

**Issue by keithrob91** [conan-io/conan#3190]
Hi,

We have CONAN_LIBS for the list of aggregated libraries and we have the specific CONAN_LIBS_XXX for the individual package libraries.

However, it would be nice to have a CONAN_LIBS_DIRECT variable that is the list of direct dependencies only.

Is that feasible?

The reason I ask is that in a package that builds a library, I would like to do:
`target_link_libraries(${LibraryName} PRIVATE ${CONAN_LIBS_DIRECT})`

and not:
`target_link_libraries(${LibraryName} PRIVATE ${CONAN_LIBS_XXX} ${CONAN_LIBS_YYY})`

or:
`target_link_libraries(${LibraryName} PRIVATE ${CONAN_LIBS})`

**Comment by danimtb** [comment]
HI @keithrob91,

CMake generator does not implement the variable you are requesting and the only way to do this right now is linking the desired libs one by one.

Could you give us more details of your use case? Why do you need only direct dependencies?

**Comment by memsharded** [comment]
Actually, using such a library list, if not using the full CONAN_LIBS variable or specifying by hand, would probably be a source of errors. Let me put an example:

- Direct dependencies of A are B & C
- B has a transitive dependency to D
- D has a transitive dependency to C
- Linking with ``target_link_libraries(mytarget ${CONAN_LIBS_DIRECT})`` will be equal to ``target_link_libraries(mytarget B C)``, and this typically cause problems with the link order.

The CMake ecosystem is moving to using targets, is there some specific reason you want to stay using old-cmake style? Are you constrained to old cmake versions?

If anything, it might make sense to return the targets that are direct requirements. CMake targets will take care of the linking order automatically.

**Comment by keithrob91** [comment]
By specifying CONAN_LIBS we force a library to have an explicit dependency on all libraries whether  direct or indirect. i.e. ldd <lib> will show them all.

We have previously not done this and only linked against the direct dependencies, leaving each library to care about what it directly needs.

Maybe this is the wrong approach?

@memsharded for RHEL6 we manually built the version of cmake that was available at the time (a number of years ago) and have not found the need to change. Additionally, wherever possible (for a number of reasons), we stick to the RHEL provided/supported rpms. For RHEL7 that is cmake 2.8.12.2 (although I do note there is an epel version that is 3.11.2).

Basically, if it isn't broke don't fix it, and being honest/blunt, we weren't aware of "modern" cmake! Having said all that, we are trying to modernise how we work, hence why we are interested in Conan, but we are a VERY small team and switching to Conan is a massive job for us and we can't change everything at once and still do what the boss wants!

**Comment by memsharded** [comment]
> We have previously not done this and only linked against the direct dependencies, leaving each library to care about what it directly needs.
Maybe this is the wrong approach?

It will fail for static linking, for example. Are you using dynamic linking only in your project? I'd say that it is more typical to list dependencies, including transitive ones, and the linker will take care of it, typically it is smart and it is able not to use libraries which symbols are finally not used.

>@memsharded for RHEL6 we manually built the version of cmake that was available at the time (a number of years ago) and have not found the need to change. Additionally, wherever possible (for a number of reasons), we stick to the RHEL provided/supported rpms. For RHEL7 that is cmake 2.8.12.2 (although I do note there is an epel version that is 3.11.2).

>Basically, if it isn't broke don't fix it, and being honest/blunt, we weren't aware of "modern" cmake! Having said all that, we are trying to modernise how we work, hence why we are interested in Conan, but we are a VERY small team and switching to Conan is a massive job for us and we can't change everything at once and still do what the boss wants!

That totally makes sense :) We know it, and this is one of the reasons we try to support the best we can down to cmake 2.8.12.

**Comment by keithrob91** [comment]
You are right, most of the libs are dynamic linking not static.

Interestingly, I've just discovered the ld argument "--as-needed". This may be part of the intelligent behaviour you describe, but is not enabled by default (on our system anyway.)

Thank you for the input, it is proving invaluable.

On the cmake 2.8.12 point, may I request that you continue this support for the foreseeable future, as removing it would be a big problem for us :-(

**Comment by memsharded** [comment]
> On the cmake 2.8.12 point, may I request that you continue this support for the foreseeable future, as removing it would be a big problem for us :-(

Yes, no problem, it is not in our roadmap to deprecate it :)
Beware, however, that it is in our roadmap to deprecate python 2 support. Not in the close future, but maybe in 2019. The main reason is that it is out of our control, some of the conan dependencies (python libraries) have already started to drop python 2 support, and that is a huge problem. So we are almost forced to follow that path (as the whole python community is finally doing). We know that this might be inconvenient for platforms like RH that use old pythons, even 2.6, but it is outside of our control, and it is doable to have a python 3 in those platforms.

> Thank you for the input, it is proving invaluable.

You are welcome, glad to help :)

Please let me know what you think would be the final conclusion. I think that having such variable might be more problematic than helpful in the mid term, and sounds like a very specific use case. Thanks!

**Comment by keithrob91** [comment]
We've decided to use CONAN_LIBS in conjunction with linker flag --as-needed.

I'll close this.
